src/main/sensors/compass.c

```diff
@@ -388,31 +404,245 @@ uint32_t compassUpdate(timeUs_t currentTimeUs)
+ *      P = ( P - Gamma(:,j) * Sw(j,:) * P );
+ *  end
+ *  zn(j) = zn(j) * lambda;
+ *  P = P / lambda;
+ *  lambda = lambda_min + (1 - lambda_min) * ( zn.' * zn ) / 3.0;
+ * 
+ * Explicit form is:
+ * 
+ *  e = [dmag(i,1) + gyro(i,3) * ( b(2) - mag(i,2) ) - gyro(i,2) * ( b(3) - mag(i,3) ); ...
+ *       dmag(i,2) - gyro(i,3) * ( b(1) - mag(i,1) ) + gyro(i,1) * ( b(3) - mag(i,3) ); ...
+ *       dmag(i,3) + gyro(i,2) * ( b(1) - mag(i,1) ) - gyro(i,1) * ( b(2) - mag(i,2) )];
+ * 
+ *  // iteration 1
+ *  dP = [P(2,1)*gyro(i,3) - P(3,1)*gyro(i,2); ...
+ *        P(2,2)*gyro(i,3) - P(3,2)*gyro(i,2); ...
+ *        P(3,2)*gyro(i,3) - P(3,3)*gyro(i,2)];
+ *  zn(1) = 1 / ( lambda + gyro(i,3)*dP(2) - gyro(i,2)*dP(3) );
+ *  g(1) = zn(1) * dP(1);
+ *  g(2) = zn(1) * dP(2);
+ *  g(3) = zn(1) * dP(3);
+ *  b(1) = b(1) - e(1) * g(1);
+ *  b(2) = b(2) - e(1) * g(2);
+ *  b(3) = b(3) - e(1) * g(3);
+ *  P(1,1) = P(1,1) - g(1) * dP(1);
+ *  P(2,1) = P(2,1) - g(2) * dP(1);
+ *  P(2,2) = P(2,2) - g(2) * dP(2);
+ *  P(3,1) = P(3,1) - g(3) * dP(1);
+ *  P(3,2) = P(3,2) - g(3) * dP(2);
+ *  P(3,3) = P(3,3) - g(3) * dP(3);
+ *
+ *  // iteration 2
+ *  dP = [P(3,1)*gyro(i,1) - P(1,1)*gyro(i,3); ...
+ *        P(3,2)*gyro(i,1) - P(2,1)*gyro(i,3); ...
+ *        P(3,3)*gyro(i,1) - P(3,1)*gyro(i,3)];
+ *  zn(2) = 1/(lambda - gyro(i,3)*dP(1) + gyro(i,1)*dP(3));
+ *  g(1) = zn(2) * dP(1);
+ *  g(2) = zn(2) * dP(2);
+ *  g(3) = zn(2) * dP(3);
+ *  b(1) = b(1) - e(2) * g(1);
+ *  b(2) = b(2) - e(2) * g(2);
+ *  b(3) = b(3) - e(2) * g(3);
+ *  P(1,1) = P(1,1) - g(1) * dP(1);
+ *  P(2,1) = P(2,1) - g(2) * dP(1);
+ *  P(2,2) = P(2,2) - g(2) * dP(2);
+ *  P(3,1) = P(3,1) - g(3) * dP(1);
+ *  P(3,2) = P(3,2) - g(3) * dP(2);
+ *  P(3,3) = P(3,3) - g(3) * dP(3);
+ *
+ *  // iteration 3
+ *  dP = [P(1,1)*gyro(i,2) - P(2,1)*gyro(i,1); ...
+ *        P(2,1)*gyro(i,2) - P(2,2)*gyro(i,1); ...
+ *        P(3,1)*gyro(i,2) - P(3,2)*gyro(i,1)];
+ *  zn(3) = 1/(lambda + gyro(i,2)*dP(1) - gyro(i,1)*dP(2));
+ *  g(1) = zn(3) * dP(1);
+ *  g(2) = zn(3) * dP(2);
+ *  g(3) = zn(3) * dP(3);
+ *  b(1) = b(1) - e(3) * g(1);
+ *  b(2) = b(2) - e(3) * g(2);
+ *  b(3) = b(3) - e(3) * g(3);
+ *  P(1,1) = P(1,1) - g(1) * dP(1);
+ *  P(2,1) = P(2,1) - g(2) * dP(1);
+ *  P(2,2) = P(2,2) - g(2) * dP(2);
+ *  P(3,1) = P(3,1) - g(3) * dP(1);
+ *  P(3,2) = P(3,2) - g(3) * dP(2);
+ *  P(3,3) = P(3,3) - g(3) * dP(3);
+
+ *  zn = zn * lambda;
+ *  P = P / lambda;
+ *  lambda = lambda_min + (1 - lambda_min) * ( zn.' * zn ) / 3.0;
+ */
+
+void compassBiasEstimatorInit(compassBiasEstimator_t *cBE, const float lambda_min, const float p0)
+{
+    compassBiasEstimatorUpdate(cBE, lambda_min, p0);
+    
+    cBE->lambda = lambda_min;
+
+    // don't know if it is necessary to initialize these as 0
+    for (unsigned i = 0; i < 3; i++) {
+        cBE->b[i] = 0.0f;
+        for (unsigned j = 0; j < 3; j++) {
+            if (i != j) {
+                cBE->P[i][j] = 0.0f;
+            }
+        }
+    }  
```

Comment:
On IEEE754 system, this will work too:
```suggestion
    memset(cBE, 0, sizeof(*cBE));     // zero contained IEEE754 floats
    cBE->lambda = lambda_min;
    compassBiasEstimatorUpdate(cBE, lambda_min, p0);
}  
```

Reply:
Cool :-), did not know that.

Reply:
Nice, many things to learn i have :-)